Create README.md (#18)
makerdao/dss-wormhole: README.md

# DAI Wormhole

> DAI Wormhole facility allows users to fast teleport DAI between "domains", i.e. different chains that have a settlement mechanism with Ethereum L1. If DAI is teleported from L2 -> L1, this is equvialent to "fast withdrawal". DAI will be minted on L1 as soon as the transaction is confirmed on L2 and sent to the user, and when transaction eventually settles on L1, DAI will be released from L1 Bridge escrow (to have DAI on L2 in the first place, it had to be put on L1 escrow some time before) and burned. 

Reviewer: ```suggestion
DAI Wormhole facility allows users to fast teleport DAI between "domains", i.e. different chains that have a settlement mechanism with Ethereum L1. 

If DAI is teleported from L2 -> L1, this is equivalent to "fast withdrawal". First, DAI will be burned on L2, then minted on L1 and sent to the user as soon as the L2 transaction is confirmed. After a while, during a settlement process, DAI will be released from L1 Bridge escrow (to have DAI on L2 in the first place, it had to be put on L1 escrow some time before) and burned. 
```

If DAI is teleported from L2 -> L2, on the source domain it will be burned and on the destination domain it will be minted, while settlement process on L1 will eventually move DAI from source domain bridge escrow to destination domain bridge escrow.

## Roles

* **Initiator** - person initiating DAI transfer by calling `initiateWormhole` . They can optionally specify Operator and Receiver 
* **Operator** - person (or specified third party) responsible for initiating minting process on destination domain by providing (in the fast path) Oracle attestations. Can call `requestMint` on `WormholeOracleAuth`

> * **Receiver** - person receiving minted DAI on a destination domain

Reviewer: Do we want to menion oracle and oracle API here?

## DAI Wormhole L2 → L1 (aka fast withdrawals) 

![FastWithdrawal](./docs/fw.png?raw=true)

### Normal (fast) path

To fast withdraw DAI from L2, user:

> * Calls `l2bridge.initiateWormhole()` - this burns DAI on L2 and sends `finalizeRegisterWormhole()` L2 -> L1 message to withdraw DAI from L2 bridge. This message, in normal cicumstances, will never be relayed and it will eventually expire in L1 message queue

Reviewer: This (same for next sections) should be ordered list

Author: Ordered, nested lists don't seem to work, leaving as it is right now

* L2 -> L1 message `finalizeFlush()` is sent to `L1Bridge` and relayed by a keeper
* `L1Bridge` upon receiving `finalizeFlush()` calls `WormholeRouter.settle()` which will
    * Transfer `DAI` from bridges' escrow to `WormholeJoin`
    * Call `WormholeJoin.settle()` which will use transfered DAI to clear any outstanding debt by calling `daiJoin.join`, `vat.frob`, `vat.slip`

### Slow (emergency) path

If attestations cannot be obtained (Oracles down or censoring), user needs to wait so that L2 message is confirmed on L1 (on Optimistic Rollups that typically is 7 days, on zkRollups it can be anything between few hours to a day). Once L2->L1 message can be relayed, user:

* Relays `finalizeRegisterWormhole()`  message to `L1Bridge`
* `L1Bridge` upon receiving `finalizeRegisterWormhole()` will call `requestMint()` on `WormholeRouter` which will:
    * Call `WormholeJoin.requestMint(wormholeGUID, maxfee)` which will
        * Check if this wormhole hasn't been used before
        * Check if the debt ceiling hasn't been reached
        * Check the current fee via `WormholeFees`
        * `vat.slip`, `vat.frob`, `daiJoin.exit`

## DAI Wormhole L2→L2

> ![Wormhole](./docs/l2.png?raw=true)

Reviewer 2: The red arrow from L2-B Bridge to WormholeJoin has a typo: `requestMin()` instead of `requestMint()`.

Author: fixed

### Normal (fast) path

Wormholing DAI to another L2 domain is very similar, the only difference is that DAI is minted on a target Domain rather then on L1. For this scheme to work MakerDAO `MCD` sytem needs to be deployed on a target domain. 

### Settlement

> Settlement process is very similar, however DAI is transfered from source domain bridge on L1 to target domain bridge on L1 before rather then moved to `L1 MCD` to pay the debt. This DAI, now in target domain bridge will be backing DAI that is minted on L2 target domain.

Reviewer 2: ```suggestion
Settlement process is very similar, however DAI is transferred from the source domain bridge on L1 to the target domain bridge on L1, rather than moved to `L1 MCD` to pay the debt. This DAI, now in the target domain bridge, will be backing DAI that is minted on the L2 target domain.
```

### Slow (emergency) path

> For a slow path, once L2->L1 message from the source domain is received on L1 and can be relayed, user can relay the message which fill call `finalizeRequestMing()` on the target domain `L1Bridge`. This will pass L1->L2 message to `L2bridge` which will call `requestMint()` on a `WormholeJoin` contract on target domain L2.

Reviewer 2: ```suggestion
For a slow path, once the L2->L1 message from the source domain is received on L1 and can be relayed, the user can relay the message, which will call `requestMint()` on the target domain `L1Bridge`. This will pass an L1->L2 message to `L2bridge` which will call `requestMint()` on a `WormholeJoin` contract on target domain L2.
```

## Technical Documenation

Each Wormhole is described with the following struct:

```
struct WormholeGUID {
	bytes32 sourceDomain;
	bytes32 targetDomain;
	address receiver;
	address operator;
	uint128 amount;
	uint80 nonce;
	uint48 timestamp;
}
```
Source domain implementation must ensure that `keccack(WorkholeGUID)` is unique for each wormhole transfer.

### Contracts

**`WormholeRouter`**

> * `file(what=="gateway", domain, gateway)` - collable only by Governance, configures gateway for a domain. If gateway is already configured, replaces it with a new one. 

Reviewer 2: ```suggestion
* `file(what=="gateway", domain, gateway)` - callable only by Governance, sets the gateway for a domain. If a gateway is already set, replaces it with a new one. 
```

> * `requestMint(WormholeGUID calldata wormholeGUID, uint256 maxFee)` - callable only by `L1Bridge`, requests `WormholeJoin` to mint DAI for the receiver of the wormhole

Reviewer 2: ```suggestion
* `requestMint(WormholeGUID calldata wormholeGUID, uint256 maxFee)` - callable only by `L1Bridge`, issues a request to mint DAI for the receiver of the wormhole. This request is made either directly to the L1 `WormholeJoin` for a fast withdrawal to L1 or indirectly by instructing the target domain's `L1Bridge` to pass an `L1 -> L2` message to the corresponding L2 `WormholeJoin` for a teleport to another L2.
```

**`WormholeJoin`**
* `requestMint(WormholeGUID calldata wormholeGUID, uint256 maxFee)` - callable either by `WormholeOracleAuth` (fast path) or by `WormholeRouter` (slow path), mints and withdraws DAI from the wormhole. If debt ceiling is reached, partial amount will be withdrawn and anything pending can be withdrawn using `withdrawPending()` later
* `withdrawPending(WormholeGUID calldata wormholeGUID, uint256 maxFee)` - callable by wormhole operator, withdraws any pending DAI from a wormhole
* `settle(bytes32 sourceDomain, uint256 batchedDaiToFlush)` - callable only by `WormhomeRouter`, settles DAI debt

**`WormholeFees`**
* `getFees(WormholeGUID calldata wormholeGUID) (uint256 fees)` - interface for getting current fee. Various implementations can be provided by the governance with different fee structures

### Authorization
* `WormholeOracleAuth`
  * `requestMint` - operator (set by the user initiating wormhole)
  * `rely`, `deny`, `file`, `addSigners`, `removeSigners` - auth (Governance)
* `WormholeRouter`

>   * `requestMint` - L1 Bridge

Reviewer 2: ```suggestion
  * `rely`, `deny`, `file` - auth (Governance)
  * `requestMint` - L1 Bridge
```
